Mono AOT Compiler packs are missing `data/UnixFilePermissions.xml`?

### Description

@davidortinau was testing .NET 8 preview 1 builds and hit the error:

```
/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/16.2.232-net8-p1/targets/Xamarin.Shared.Sdk.targets(1037,3): error MSB4018: The "AOTCompile" task failed unexpectedly.
System.AggregateException: One or more errors occurred. (One or more errors occurred. (An error occurred trying to start process '/usr/local/share/dotnet/packs/Microsoft.NETCore.App.Runtime.AOT.osx-x64.Cross.maccatalyst-x64/8.0.0-preview.1.23106.5/Sdk/../tools/mono-aot-cross' with working directory '/Users/dmo/work/ControlGallery/src/ControlGallery'. Permission denied))
 ---> System.AggregateException: One or more errors occurred. (An error occurred trying to start process '/usr/local/share/dotnet/packs/Microsoft.NETCore.App.Runtime.AOT.osx-x64.Cross.maccatalyst-x64/8.0.0-preview.1.23106.5/Sdk/../tools/mono-aot-cross' with working directory '/Users/dmo/work/ControlGallery/src/ControlGallery'. Permission denied)
 ---> System.ComponentModel.Win32Exception (13): An error occurred trying to start process '/usr/local/share/dotnet/packs/Microsoft.NETCore.App.Runtime.AOT.osx-x64.Cross.maccatalyst-x64/8.0.0-preview.1.23106.5/Sdk/../tools/mono-aot-cross' with working directory '/Users/dmo/work/ControlGallery/src/ControlGallery'. Permission denied
   at System.Diagnostics.Process.ForkAndExecProcess(ProcessStartInfo startInfo, String resolvedFilename, String[] argv, String[] envp, String cwd, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at Xamarin.Utils.Execution.<>c__DisplayClass24_1.<RunAsync>b__0() in /Users/builder/azdo/_work/1/s/xamarin-macios/tools/common/Execution.cs:line 103
--- End of stack trace from previous location ---
   at Xamarin.MacDev.Tasks.XamarinTask.ExecuteAsync(String fileName, IList`1 arguments, String sdkDevPath, Dictionary`2 environment, Boolean mergeOutput, Boolean showErrorIfFailure, String workingDirectory) in /Users/builder/azdo/_work/1/s/xamarin-macios/msbuild/Xamarin.MacDev.Tasks/Tasks/XamarinTask.cs:line 120
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Xamarin.MacDev.Tasks.AOTCompileTaskBase.<>c__DisplayClass36_0.<Execute>b__7(Task`1 v) in /Users/builder/azdo/_work/1/s/xamarin-macios/msbuild/Xamarin.MacDev.Tasks/Tasks/AOTCompileTaskBase.cs:line 108
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.WaitAllCore(Task[] tasks, Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.WaitAll(Task[] tasks)
   at Xamarin.MacDev.Tasks.AOTCompileTaskBase.Execute() in /Users/builder/azdo/_work/1/s/xamarin-macios/msbuild/Xamarin.MacDev.Tasks/Tasks/AOTCompileTaskBase.cs:line 115
   at Xamarin.MacDev.Tasks.AOTCompile.Execute() in /Users/builder/azdo/_work/1/s/xamarin-macios/msbuild/Xamarin.MacDev.Tasks/Tasks/AOTCompile.cs:line 14
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/Users/dmo/work/ControlGallery/src/ControlGallery/ControlGallery.csproj]
```

We found that running this fixed it:

```
chmod +x /usr/local/share/dotnet/packs/Microsoft.NETCore.App.Runtime.AOT.osx-x64.Cross.maccatalyst-x64/8.0.0-preview.1.23106.5/tools/mono-aot-cross
```

The Android/iOS workloads define a `data/UnixFilePermission.xml` file, so that `dotnet workload` commands can supply the execute bit. Details at: https://github.com/dotnet/sdk/issues/16894

We think this was implemented sometime in dotnet/runtime: https://github.com/dotnet/runtime/pull/54501

But I don't see the file here: 

https://dev.azure.com/dnceng/public/_artifacts/feed/dotnet8/NuGet/Microsoft.NETCore.App.Runtime.AOT.osx-x64.Cross.maccatalyst-x64/overview/8.0.0-preview.1.23106.5

### Reproduction Steps

* Install .NET 8 Preview 1 MAUI bits: https://gist.github.com/jonathanpeppers/99961b4639823206c2537c03c856b097
* `dotnet new maccatalyst`
* `dotnet build`


### Expected behavior

You should *not* need to `chmod +x mono-aot-cross`

### Actual behavior

Some Mac machines you have to `chmod +x mono-aot-cross`.

### Regression?

Yes

### Known Workarounds

```
chmod +x /usr/local/share/dotnet/packs/Microsoft.NETCore.App.Runtime.AOT.osx-x64.Cross.maccatalyst-x64/8.0.0-preview.1.23106.5/tools/mono-aot-cross
```

### Configuration

.NET SDK 8.0.100-preview.1.23109.10

### Other information

_No response_

--- a/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.UnixFilePermissions.xml.in
+++ b/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.UnixFilePermissions.xml.in
@@ -1,0 +1,3 @@
+<FileList>
+    ${PermissionsProperties}
+</FileList>

--- a/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj
+++ b/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.MonoCrossAOT.sfxproj
@@ -19,6 +19,7 @@
 
   <ItemGroup>
     <NativeRuntimeAsset Include="$(IntermediateOutputPath)$(TargetCrossRid).Sdk.props" TargetPath="Sdk/Sdk.props" />
+    <NativeRuntimeAsset Condition="!$([MSBuild]::IsOsPlatform('Windows'))" Include="$(IntermediateOutputPath)Microsoft.NETCore.App.MonoCrossAOT.UnixFilePermissions.xml" TargetPath="data/UnixFilePermissions.xml" />
   </ItemGroup>
 
   <Target Name="WriteTemplateFiles" BeforeTargets="ValidateProperties">
@@ -38,7 +39,15 @@
     <PropertyGroup>
       <_PermissionsFiles>@(_ToolFile -> '&lt;File Path=&quot;tools/%(RecursiveDir)%(FileName)%(Extension)&quot; Permission=&quot;755&quot; /&gt;', ' ')</_PermissionsFiles>
     </PropertyGroup>
+    <ItemGroup>
+      <_PermissionsProperties Include="PermissionsProperties" Value="$(_PermissionsFiles)" />
+    </ItemGroup>
 
+    <GenerateFileFromTemplate
+      Condition="!$([MSBuild]::IsOsPlatform('Windows'))"
+      TemplateFile="Microsoft.NETCore.App.MonoCrossAOT.UnixFilePermissions.xml.in"
+      Properties="@(_PermissionsProperties)"
+      OutputPath="$(IntermediateOutputPath)Microsoft.NETCore.App.MonoCrossAOT.UnixFilePermissions.xml" />
     <GenerateFileFromTemplate
       TemplateFile="Microsoft.NETCore.App.MonoCrossAOT.Sdk.props.in"
       Properties="@(_SdkPropsProperties)"

--- a/src/mono/wasm/Wasm.Build.Tests/WorkloadTests.cs
+++ b/src/mono/wasm/Wasm.Build.Tests/WorkloadTests.cs
@@ -1,0 +1,119 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Xml;
+using System.Xml.Serialization;
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+#nullable enable
+
+namespace Wasm.Build.Tests
+{
+    public class WorkloadTests : BuildTestBase
+    {
+        public WorkloadTests(ITestOutputHelper output, SharedBuildPerTestClassFixture buildContext)
+            : base(output, buildContext)
+        {
+        }
+
+        [Fact]
+        [SkipOnPlatform(TestPlatforms.Windows, "Not applicable on windows")]
+        public void FilesInUnixFilesPermissionsXmlExist()
+        {
+            // not doing any project generation here
+            _enablePerTestCleanup = false;
+
+            // find all the UnixFilePermissions ..
+            string packsDir = Path.Combine(Path.GetDirectoryName(s_buildEnv.DotNet)!, "packs");
+            Assert.True(Directory.Exists(packsDir), $"Could not find packs directory {packsDir}");
+
+            var unixPermFiles = Directory.EnumerateFiles(packsDir, "UnixFilePermissions.xml", new EnumerationOptions { RecurseSubdirectories = true });
+            foreach (string unixPermFile in unixPermFiles)
+            {
+                Assert.True(File.Exists(unixPermFile), $"Could not find {unixPermFile}");
+                FileList? list = FileList.Deserialize(unixPermFile);
+                if (list == null)
+                    throw new Exception($"Could not read unix permissions file {unixPermFile}");
+
+                // File is in <packs>/<packName>/<version>/data/UnixFilePermissions.xml
+                // and <FileList><File Path="tools/bin/2to3" Permission="755" />
+                string thisPackDir = Path.Combine(Path.GetDirectoryName(unixPermFile)!, "..");
+                foreach (FileListFile flf in list.File)
+                {
+                    if (flf.Path == null)
+                        throw new Exception($"Path for FileListFile should not be null. xml: {unixPermFile}");
+
+                    var targetFile = Path.Combine(thisPackDir, flf.Path);
+                    Assert.True(File.Exists(targetFile), $"Expected file {targetFile} to exist in the pack, as it is referenced in {unixPermFile}");
+                }
+            }
+
+            // We don't install the cross compiler pack from nupkg, so we don't
+            // have the unixFilePermissions for that
+            // Expect just the emscripten ones here for now
+
+            // linux doesn't have Emscripten.Python package, so only 2 there
+            int expectedPermFileCount = RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ? 3 : 4;
+
+            int permFileCount = unixPermFiles.Count();
+            if (permFileCount != expectedPermFileCount)
+                throw new XunitException($"Expected to find 3 UnixFilePermissions.xml files, but got {permFileCount}."
+                                            + $"{Environment.NewLine}Files: {string.Join(", ", unixPermFiles)}");
+        }
+    }
+
+    [Serializable]
+    [XmlType(AnonymousType = true)]
+    [XmlRoot(Namespace = "", IsNullable = false)]
+    public class FileList
+    {
+        private FileListFile[]? fileField;
+
+        [XmlElement("File")]
+        public FileListFile[] File
+        {
+            get => fileField ?? Array.Empty<FileListFile>();
+            set => fileField = value;
+        }
+
+        public static FileList? Deserialize(string pathToXml)
+        {
+            var serializer = new XmlSerializer(typeof(FileList));
+
+            using var fs = new FileStream(pathToXml, FileMode.Open, FileAccess.Read);
+            var reader = XmlReader.Create(fs);
+            FileList? fileList = (FileList?)serializer.Deserialize(reader);
+            return fileList;
+        }
+    }
+
+    // From https://github.com/dotnet/sdk/blob/main/src/Cli/dotnet/NugetPackageDownloader/WorkloadUnixFilePermissionsFileList.cs
+    [Serializable]
+    [XmlType(AnonymousType = true)]
+    public class FileListFile
+    {
+        private string? pathField;
+
+        private string? permissionField;
+
+        [XmlAttribute]
+        public string? Path
+        {
+            get => pathField;
+            set => pathField = value;
+        }
+
+        [XmlAttribute]
+        public string? Permission
+        {
+            get => permissionField;
+            set => permissionField = value;
+        }
+    }
+}

